File: platform/macos/Settings/NativeDocumentExport.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation

public enum NativeDocumentExportLoss: String, CaseIterable, Hashable, Sendable {
  case failureAlerts, remoteResize, networkFamilies, pointerTiming, clipboardLimit, windowPlacement, displayIdentity, ignoredInput, sshGateway
  public var description: String {
    switch self {
    case .failureAlerts: String(localized:"document.failure.alerts.omitted", defaultValue:"Failure-alert settings are not supported by this connection-file format. The receiving viewer will use its own error-alert policy.")
    case .remoteResize: String(localized:"document.remote.resize.settings.are.not.supported.by.this.connection.file.format.the", defaultValue:"Remote-resize settings are not supported by this connection-file format. The receiving viewer will use its own settings.")
    case .networkFamilies: String(localized:"document.ipv4.and.ipv6.settings.are.not.supported.by.this.connection.file.format", defaultValue:"IPv4 and IPv6 settings are not supported by this connection-file format. The receiving viewer will use its own IP version settings.")
    case .pointerTiming: String(localized:"document.pointer.event.timing.is.not.supported.by.this.connection.file.format.the", defaultValue:"Pointer-event timing is not supported by this connection-file format. The receiving viewer will use its own pointer timing.")
    case .clipboardLimit: String(localized:"document.the.incoming.clipboard.size.limit.is.not.supported.by.this.connection.file", defaultValue:"The incoming clipboard size limit is not supported by this connection-file format. The receiving viewer will use its own limit.")
    case .windowPlacement: String(localized:"document.initial.window.size.position.and.maximization.are.not.supported.by.this.connection", defaultValue:"Initial window size, position and maximization are not supported by this connection-file format. The receiving viewer will use its own window settings.")
    case .displayIdentity: String(localized:"document.stable.display.identities.become.the.monitor.numbers.listed.in.this.review.the", defaultValue:"Stable display identities become the monitor numbers listed in this review. The receiving viewer interprets those numbers using its own display arrangement.")
    case .sshGateway: String(localized:"document.the.ssh.gateway.cannot.be.saved.in.this.connection.file.format.opening", defaultValue:"The SSH gateway cannot be saved in this connection-file format. Opening the exported file will connect directly unless you configure the SSH gateway separately.")
    case .ignoredInput: String(localized:"document.fields.ignored.when.opening.the.original.file.will.not.be.copied.to", defaultValue:"Fields ignored when opening the original file will not be copied to the exported file.")
    }
  }
}
public enum NativeDocumentExportError: Error, Equatable, Sendable, CustomStringConvertible {
  case reviewRequired, securityPolicy, displayMapping, invalidConfiguration, unavailable
  public var description: String {
    switch self {
    case .reviewRequired: String(localized:"document.review.the.settings.that.cannot.be.preserved.before.exporting", defaultValue:"Review the settings that cannot be preserved before exporting.")
    case .securityPolicy: String(localized:"document.this.file.format.cannot.preserve.the.custom.tls.priority.policy.use.a", defaultValue:"This file format cannot preserve the custom TLS priority policy. Use a native profile to retain it.")
    case .displayMapping: String(localized:"document.the.selected.displays.cannot.be.mapped.to.current.monitor.numbers.resolve.the", defaultValue:"The selected displays cannot be mapped to current monitor numbers. Resolve the display selection before exporting.")
    case .invalidConfiguration: String(localized:"document.a.connection.setting.cannot.be.represented.in.this.file.correct.it.before", defaultValue:"A connection setting cannot be represented in this file. Correct it before exporting.")
    case .unavailable: String(localized:"document.finish.the.current.connection.operation.or.settings.edit.before.exporting", defaultValue:"Finish the current connection operation or settings edit before exporting.")
    }
  }
}

// Immutable, preflighted non-secret compatibility output. Construction and review
// never open a file or mutate the supplied settings. No credential/trust-store
// object, invocation environment, unknown raw entry or arbitrary key is accepted.
public struct NativeDocumentExport: Sendable, Identifiable {
  public let id = UUID()
  public let losses: Set<NativeDocumentExportLoss>
  public let endpoint: String
  public let monitorIndices: [NativeDisplayID:Int]
  private let data: Data
  public func serializedData(acknowledging losses: Set<NativeDocumentExportLoss> = []) throws -> Data {
    guard self.losses.isSubset(of:losses) else { throw NativeDocumentExportError.reviewRequired }
    return data
  }
  public init(endpoint: String, configuration: NativeSessionConfiguration,
              inactiveCursor: NativeCursorFallback = .dot,
              legacyDisplays: [NativeDisplayID] = [], ignoredInput: Bool = false,
              monitorIndices: [NativeDisplayID:Int]? = nil, sshGateway: NativeSSHGateway? = nil) throws {
    guard configuration.tlsPriority.isEmpty else { throw NativeDocumentExportError.securityPolicy }
    guard inactiveCursor == .dot || inactiveCursor == .system,
          configuration.input.shortcutModifiers.rawValue & ~15 == 0 else { throw NativeDocumentExportError.invalidConfiguration }
    if !endpoint.isEmpty {
      do { try NativeEndpoint.validate(endpoint) }
      catch { throw NativeDocumentExportError.invalidConfiguration }
    }
    guard NativeTrustFiles.isValidPath(configuration.caFile), NativeTrustFiles.isValidPath(configuration.crlFile) else {
      throw NativeDocumentExportError.invalidConfiguration
    }
    var assignments: [NativeDocumentAssignment] = [.init("ServerName",endpoint)]
    func add(_ name: String,_ value: String) { assignments.append(.init(name,value)) }
    func flag(_ name: String,_ value: Bool) { add(name,value ? "on" : "off") }
    let choices = try NativeSecuritySelection.choices()
    let types: String
    if let requested = configuration.securityTypes {
      let names = try requested.map { id in
        guard let choice = choices.first(where:{ $0.id == id }), choice.available else { throw NativeDocumentExportError.invalidConfiguration }
        return choice.name
      }
      types = try NativeSecuritySelection(names.joined(separator:",")).canonical
    } else { types = try NativeSecuritySelection().canonical }
    add("SecurityTypes",types); add("X509CA",configuration.caFile); add("X509CRL",configuration.crlFile)
    flag("Shared",configuration.shared); flag("ReconnectOnError",configuration.reconnectOnError)
    flag("SendClipboard",configuration.clipboardSend); flag("AcceptClipboard",configuration.clipboardReceive)
    let encoding = try configuration.encoding ?? NativeEncodingOptions()
    for field in try NativeEncodingOptions.schema() where field.persistent { add(field.name,try encoding.value(for:field.id).value) }
    let input = configuration.input
    flag("ViewOnly",input.viewOnly); flag("EmulateMiddleButton",input.emulateMiddle)
    flag("FullscreenSystemKeys",input.fullscreenSystemKeys)
    let modifiers = [(UInt32(1),"Ctrl"),(2,"Shift"),(4,"Alt"),(8,"Super")]
      .filter { input.shortcutModifiers.rawValue & $0.0 != 0 }.map(\.1).joined(separator:",")
    add("ShortcutModifiers",modifiers)
    flag("AlwaysCursor",input.cursorFallback != .hidden)
    let shape = input.cursorFallback == .hidden ? inactiveCursor : input.cursorFallback
    add("CursorType",shape == .system ? "System" : "Dot")
    let scaling = configuration.scaling ?? .builtIn
    add("ScalingFactor",scaling.canonical)
    add("ScalingQuality",scaling.filter == .nearest ? "Nearest" : scaling.filter == .area ? "Area" : "Bilinear")
    add("DesktopPixelUnits",scaling.devicePixels ? "Device" : "Logical")
    let fullscreen = configuration.fullscreenPolicy
    flag("FullScreen",fullscreen.startsFullscreen)
    add("FullScreenMode",fullscreen.mode == .selected ? "Selected" : fullscreen.mode == .all ? "All" : "Current")
    // Even built-in omitted values can differ from the receiving viewer’s
    // preferences, so omission requires review for every export.
    var losses: Set<NativeDocumentExportLoss> = [.failureAlerts,.remoteResize,.networkFamilies,.pointerTiming,.clipboardLimit,.windowPlacement]
    var indices: [Int] = []
    var mapping: [NativeDisplayID:Int] = [:]
    if let monitorIndices {
      guard Set(monitorIndices.keys) == Set(fullscreen.selectedDisplays),
            Set(monitorIndices.values).count == monitorIndices.count,
            monitorIndices.values.allSatisfy({ $0 > 0 && $0 <= Int(Int32.max) }) else { throw NativeDocumentExportError.displayMapping }
      mapping = monitorIndices
    }
    if !fullscreen.selectedDisplays.isEmpty {
      if monitorIndices == nil {
        guard legacyDisplays.count <= 64, Set(legacyDisplays).count == legacyDisplays.count else { throw NativeDocumentExportError.displayMapping }
        for id in fullscreen.selectedDisplays {
          guard let index = legacyDisplays.firstIndex(of:id) else { throw NativeDocumentExportError.displayMapping }
          mapping[id] = index+1
        }
      }
      indices = Array(mapping.values)
      losses.insert(.displayIdentity)
    }
    add("FullScreenSelectedMonitors",indices.sorted().map(String.init).joined(separator:","))
    if ignoredInput { losses.insert(.ignoredInput) }
    if sshGateway != nil { losses.insert(.sshGateway) }
    let bytes = try NativeConnectionDocument.serialize(assignments)
    // Verify every emitted field through the shared file-semantic boundary, not
    // just syntax. A future schema field absent from the file catalog fails.
    let parsed = try NativeConnectionDocument(data:bytes)
    for index in parsed.entries.indices {
      guard try parsed.validatedOption(at:index) != nil else { throw NativeDocumentExportError.invalidConfiguration }
    }
    self.endpoint = endpoint; self.losses = losses; self.monitorIndices = mapping; data = bytes
  }
}

File: platform/windows/TidyVNC.Native/Documents/NativeDocumentExport.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using TidyVNC.Native.Trust;

namespace TidyVNC.Native.Documents;

public enum NativeExportProblem
{
    /// <summary>A custom TLS priority cannot be carried by the file format; a native profile keeps it.</summary>
    SecurityPolicy,
    /// <summary>Selected displays have no monitor number in the current arrangement.</summary>
    DisplayMapping,
    /// <summary>A setting cannot be represented in the file.</summary>
    InvalidConfiguration,
    /// <summary>Losses were not acknowledged.</summary>
    ReviewRequired,
}

public sealed class NativeExportFailure(NativeExportProblem problem) : Exception($"Connection export: {problem}")
{
    public NativeExportProblem Problem { get; } = problem;
}

/// <summary>A value snapshot of one connection's settings, taken before any UI appears.</summary>
public sealed record NativeExportSource(
    string Endpoint, bool Shared, bool ReconnectOnError, bool ClipboardSend, bool ClipboardReceive, NativeEncodingOptions Encoding,
    NativeInputSettings Input, NativeCursorFallback InactiveCursor, NativeScaling Scaling, NativeFullscreenPolicy Fullscreen,
    string SecurityTypes, string TlsPriority, string CaFile, string CrlFile, bool IgnoredInput, bool SshGateway);

/// <summary>
/// Save connection file as (macOS NativeDocumentExport; PARITY F05-F07): a
/// preflighted, non-secret compatibility file of the current settings. The
/// core decides the losses the review must show (W2 export-loss rules); a
/// custom TLS priority refuses the export; stable display IDs become the
/// monitor numbers of the current arrangement. Every emitted field is read
/// back through the shared file decoder before the export is offered.
/// Passwords and trust decisions are never written.
/// </summary>
public sealed class NativeDocumentExport
{
    private readonly byte[] data;

    public NativeExportLoss Losses { get; }
    public string Endpoint { get; }
    /// <summary>Display ID to exported monitor number.</summary>
    public IReadOnlyDictionary<string, int> MonitorIndices { get; }

    private NativeDocumentExport(byte[] data, NativeExportLoss losses, string endpoint, IReadOnlyDictionary<string, int> indices)
    {
        this.data = data; Losses = losses; Endpoint = endpoint; MonitorIndices = indices;
    }

    /// <summary>The file, only once every loss has been acknowledged.</summary>
    public byte[] Data(NativeExportLoss acknowledged) =>
        (Losses & ~acknowledged) == 0 ? data : throw new NativeExportFailure(NativeExportProblem.ReviewRequired);

    /// <param name="legacyDisplays">Connected displays in the retained monitor numbering.</param>
    /// <param name="monitorIndices">
    /// Numbers the user chose for every selected display (<see cref="NativeExportMapping"/>),
    /// or null to number them from the current arrangement.
    /// </param>
    public static NativeDocumentExport Create(NativeExportSource source, IReadOnlyList<string> legacyDisplays,
                                              IReadOnlyDictionary<string, int>? monitorIndices = null)
    {
        NativeExportLoss losses;
        try
        {
            losses = NativeExportLosses.For(!source.Fullscreen.SelectedDisplays.IsEmpty, source.IgnoredInput, source.SshGateway, source.TlsPriority);
        }
        catch (NativeExportRefused) { throw new NativeExportFailure(NativeExportProblem.SecurityPolicy); }
        if (source.InactiveCursor is not (NativeCursorFallback.Dot or NativeCursorFallback.System) ||
            ((uint)source.Input.ShortcutModifiers & ~15u) != 0 ||
            (source.CaFile.Length != 0 && !NativeTrustFiles.IsValidPath(source.CaFile)) ||
            (source.CrlFile.Length != 0 && !NativeTrustFiles.IsValidPath(source.CrlFile)))
            throw new NativeExportFailure(NativeExportProblem.InvalidConfiguration);
        if (source.Endpoint.Length != 0)
        {
            try { NativeEndpoint.Validate(source.Endpoint); }
            catch (NativeError) { throw new NativeExportFailure(NativeExportProblem.InvalidConfiguration); }
        }

        var fields = new List<NativeDocumentAssignment>();
        void Add(string name, string value) => fields.Add(new NativeDocumentAssignment(name, value));
        void Flag(string name, bool value) => Add(name, value ? "on" : "off");
        Add("ServerName", source.Endpoint);
        Add("SecurityTypes", source.SecurityTypes);
        Add("X509CA", source.CaFile); Add("X509CRL", source.CrlFile);
        Flag("Shared", source.Shared); Flag("ReconnectOnError", source.ReconnectOnError);
        Flag("SendClipboard", source.ClipboardSend); Flag("AcceptClipboard", source.ClipboardReceive);
        foreach (var field in NativeEncodingOptions.Schema().Where(f => f.Persistent)) Add(field.Name, source.Encoding.Value(field.Id).Value);
        var input = source.Input;
        Flag("ViewOnly", input.ViewOnly); Flag("EmulateMiddleButton", input.EmulateMiddle); Flag("FullscreenSystemKeys", input.FullscreenSystemKeys);
        Add("ShortcutModifiers", NativeInputSettings.Canonical(input.ShortcutModifiers));
        Flag("AlwaysCursor", input.CursorFallback != NativeCursorFallback.Hidden);
        var shape = input.CursorFallback == NativeCursorFallback.Hidden ? source.InactiveCursor : input.CursorFallback;
        Add("CursorType", shape == NativeCursorFallback.System ? "System" : "Dot");
        Add("ScalingFactor", source.Scaling.Canonical); Add("ScalingQuality", source.Scaling.Filter.Canonical());
        Add("DesktopPixelUnits", source.Scaling.DevicePixels ? "Device" : "Logical");
        var fullscreen = source.Fullscreen;
        Flag("FullScreen", fullscreen.StartsFullscreen);
        Add("FullScreenMode", NativeFullscreenPolicy.Canonical(fullscreen.Mode));
        var indices = new Dictionary<string, int>(StringComparer.Ordinal);
        foreach (var id in fullscreen.SelectedDisplays)
        {
            var index = monitorIndices is null ? legacyDisplays.ToList().IndexOf(id) + 1 : monitorIndices.GetValueOrDefault(id);
            if (index <= 0) throw new NativeExportFailure(NativeExportProblem.DisplayMapping);
            indices[id] = index;
        }
        if (monitorIndices is not null && (monitorIndices.Count != indices.Count || indices.Values.Distinct().Count() != indices.Count))
            throw new NativeExportFailure(NativeExportProblem.DisplayMapping);
        Add("FullScreenSelectedMonitors", string.Join(',', indices.Values.Order()));

        byte[] bytes;
        try
        {
            bytes = NativeConnectionDocument.Serialize(fields);
            // Every emitted field must pass the shared file semantics, not only the syntax.
            using var parsed = new NativeConnectionDocument(bytes);
            for (var i = 0; i < parsed.Entries.Count; i++)
                if (parsed.ValidatedOption(i) is null) throw new NativeExportFailure(NativeExportProblem.InvalidConfiguration);
        }
        catch (Exception error) when (error is NativeDocumentFailure or NativeError)
        {
            throw new NativeExportFailure(NativeExportProblem.InvalidConfiguration);
        }
        return new NativeDocumentExport(bytes, losses, source.Endpoint, indices);
    }
}

/// <summary>
/// Exported monitor numbers chosen by the user (macOS NativeDocumentExportMapping):
/// offered when a selected display has no number in the current arrangement,
/// or to renumber for the receiving computer. Only the file changes; the
/// connection keeps its selected displays.
/// </summary>
public sealed class NativeExportMapping
{
    public Guid Id { get; } = Guid.NewGuid();
    public IReadOnlyList<string> SelectedDisplays { get; }
    public IReadOnlyDictionary<string, int> Suggested { get; }

    public NativeExportMapping(NativeFullscreenPolicy policy, IReadOnlyList<string> legacyDisplays, IReadOnlyDictionary<string, int>? previous = null)
    {
        SelectedDisplays = [.. policy.SelectedDisplays];
        if (previous is not null) { Suggested = previous; return; }
        var suggested = new Dictionary<string, int>(StringComparer.Ordinal);
        if (legacyDisplays.Count <= 64 && legacyDisplays.Distinct(StringComparer.Ordinal).Count() == legacyDisplays.Count)
            foreach (var id in SelectedDisplays)
                if (legacyDisplays.ToList().IndexOf(id) is var index and >= 0) suggested[id] = index + 1;
        Suggested = suggested;
    }

    /// <summary>Distinct positive whole numbers for every selected display, or null.</summary>
    public IReadOnlyDictionary<string, int>? Indices(IReadOnlyDictionary<string, string> text)
    {
        var result = new Dictionary<string, int>(StringComparer.Ordinal);
        foreach (var id in SelectedDisplays)
        {
            if (!text.TryGetValue(id, out var value)) return null;
            var input = value.Trim();
            if (input.Length is 0 or > 10 || !input.All(char.IsAsciiDigit) || !int.TryParse(input, out var number) || number <= 0) return null;
            result[id] = number;
        }
        return result.Values.Distinct().Count() == result.Count ? result : null;
    }
}
